rfcs/0000-import-expression-ast-node.md

```diff
@@ -0,0 +1,204 @@
+- Repo: `babel/babel`
+- Start Date: 2020-05-01
+- RFC PR: https://github.com/babel/rfcs/pull/4
+- Related Issues: https://github.com/babel/babel/pull/10962
+- Authors: @JLHwung
+- Champion: @JLHwung
+- Implementors: @JLHwung
+
+# Summary
+
+This RFC proposes to change the AST shape of dynamic import `import(moduleName)` in ES2020. A new `ImportExpression` AST node is introduced.
+
+# Motivation
+
+Currently dynamic import is parsed as a `CallExpression`, whose `callee` is a pseudo expression node `Import`. For example, here is the parsed AST of `import("foo")`.
+
+```json
+{
+  "type": "CallExpression",
+  "callee": {
+    "type": "Import"
+  },
+  "arguments": [
+    {
+      "type": "StringLiteral",
+      "value": "foo"
+    }
+  ]
+}
+```
+
+It encourages an incorrect mental model, as developers who may think of `import()` as a function call. It is also problematic when we extend the current AST shape to support stage-1 [module attributes](https://github.com/tc39/proposal-module-attributes) (`import(moduleName, attributes)`) since the arguments (`moduleName` and `attributes` respectively) share the same semantics in the AST.
+
+# Detailed design
+
+Introduce a new `ImportExpression` AST node type in `@babel/types`.
+
+```js
+interface ImportExpression <: Expression {
+  type: "ImportExpression";
+  source: Expression;
+}
+```
+
+The example above will now be parsed as
+
+```json
+{
+  "type": "ImportExpression",
+  "source": {
+    "type": "StringLiteral",
+    "value": "foo"
+  }
+}
+```
+
+## Revision on the `CallExpression` interface
+
+`Import` will be removed from the allowed `callee` type of the `CallExpression` node.
+
+```diff
+interface CallExpression <: Expression {
+- callee: Expression | Super | Import
++ callee: Expression | Super
+}
+```
+
+## Matching `import()`
+Currently a plugin can test if a node is an `import()` by visiting `CallExpression` and filtering by the `callee`.
+
+```js
+// Babel 7
+export default {
+  name: "my-plugin",
+  visitor: {
+    CallExpression({ node }) {
+      if (node.callee.type === "Import") {
+        const source = node.arguments[0];
+      }
+    }
+  }
+}
+```
+
+This RFC proposes `import()` to be parsed a dedicated AST type. Now plugin authors can match `import()` by `ImportExpression`.
+
+```js
+// Babel 8
+export default {
+  name: "my-plugin",
+  visitor: {
+    ImportExpression({ node }) {
+      const source = node.source;
+    }
+  }
+}
+```
+
+A plugin can also support both Babel 7 and Babel 8.
+```js
+// Babel 8
+export default {
+  name: "my-plugin",
+  visitor: {
+    ImportExpression({ node }) {
+      const source = node.source;
+    },
+    // Compatibility for Babel 7
+    CallExpression({ node }) {
+      if (node.callee.type === "Import") {
+        const source = node.arguments[0];
+      }
+    }
+  }
+}
+```
```

Comment:
This won't work. If you try to run this code with Babel 7 (or with Babel 6), it throws an error:
```js
/home/nicolo/Documenti/dev/babel-plugin-dynamic-import-node/node_modules/babel-traverse/lib/visitors.js:196
      throw new Error(messages.get("traverseVerifyNodeType", nodeType));
      ^

Error: You gave us a visitor for the node type "ImportExpression" but it's not a valid type
    at verify (/home/nicolo/Documenti/dev/babel-plugin-dynamic-import-node/node_modules/babel-traverse/lib/visitors.js:196:13)
    at Function.explode (/home/nicolo/Documenti/dev/babel-plugin-dynamic-import-node/node_modules/babel-traverse/lib/visitors.js:72:3)
```

A plugin with support for both Babel 7 and Babel 8 would look like this:
```js
export default function (api) {
  const version = Number(api.version.match(/^\d+/)[0]);

  if (version >= 8) {
    return {
      visitor: {
        ImportExpression({ node }) {
          const source = node.source;
        },
      },
    };
  }

  return {
    CallExpression(path) {
      if (node.callee.type === "Import") {
        const source = node.arguments[0];
      }
    },
  };
}
```

Reply:
Oh I thought we don't validate unknown type at all. Will update the example.

Reply:
Updated the example. The version detection is replaced by brand detection for aesthetic reasons.